### PR TITLE
Initial implementation.

### PR DESCRIPTION
**Pull request by timlesallen**
To test with staging:
```sh
DATABASE_URL=`heroku config:get DATABASE_URL -a cs-media-search-logger-staging`?ssl=1 \
  DATABASE_TABLE_NAME=`heroku config:get DATABASE_TABLE_NAME -a cs-media-search-logger-staging` \
  AMQP_URL=`heroku config:get AMQP_URL -a cs-image-crawl-staging` \
  AMQP_SEARCH_QUEUE=`heroku config:get AMQP_SEARCH_QUEUE -a cs-image-crawl-staging` \
  npm start
```

And when we come to running this for real (replay all logs since 2018-12-12):
```sh
DATABASE_URL=`heroku config:get RDS_DATABASE_URL -a cs-media-search-logger` \
  DATABASE_TABLE_NAME=`heroku config:get DATABASE_TABLE_NAME -a cs-media-search-logger` \
  AMQP_URL=`heroku config:get AMQP_URL -a cs-image-crawl` \
  AMQP_SEARCH_QUEUE=`heroku config:get AMQP_SEARCH_QUEUE -a cs-image-crawl` \
  LOG_OFFSET=10525620 \
  npm start
```

Note: We will want to do this from either an EC2 instance or a heroku box to get fast up/down speeds between the DB and AMQP.

### CODE REVIEW

File: src/amqp-writable-stream.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+const { Writable } = require('stream');
```
Comment (eugeneware) [inline src/amqp-writable-stream.js:1]: this could probably be its own module

File: src/_index.xs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+// TODO: cs-media-search-logger-staging and cs-image-crawl-staging use differend DBs...
+// they should probably both use the cs-media-search-logger-staging one like we do in live.
+//
+// DATABASE_URL=postgres://u5e2ab49ouu1tv:pe8177b072fbee7da07644814bc3e8b4fb82b9a2f05bc52b6afea876611ccc905@ec2-35-153-215-128.compute-1.amazonaws.com:5432/d2sl32lqah04e2
+// DATABASE_TABLE_NAME=image-search-logs
+
+const { Writable } = require('stream');
+
+// this only makes sure we don't write too much at once... it doesn't stop us from overloading the rabbitmq server... :(
+class AmqpWritableStream extends Writable {
```
Comment (eugeneware) [inline src/_index.xs]: This might also live in its own module.

File: src/_index.xs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+//
+// Note: a highWaterMark of 16 with a batchSize of 100 is going to be the default behaviour
+// @see https://nodejs.org/api/stream.html#stream_new_stream_readable_options for readable
+// stream options (specifically the default values).
+
+// const amqp = require('amqplib');
+let conn;
+
+// TODO: cs-media-search-logger-staging and cs-image-crawl-staging use differend DBs...
+// they should probably both use the cs-media-search-logger-staging one like we do in live.
```
Comment (timlesallen) [inline src/_index.xs]: My intention has been that there was no real reason to have them in the same DB - we should be able to split the live stuff if we needed to.

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+const knexLog = require('knex-log')(knex, { tableName: config.tableName, columnName: 'data' });
+const amqp = require('amqplib');
+const AmqpWritableStream = require('./amqp-writable-stream');
+const progress = require('progress-stream');
+const { LogFrame, CompositeLogView, RawLogView } = require('log-frame');
+const { ProgressBar } = require('logf-progress');
+
+main();
+
+async function main () {
```
Comment (timlesallen) [inline src/index.js:10]: main function is quite long.  Maybe bring the progress stuff out at least?

**Comment by timlesallen** [comment]
How does backpressure work?  Did you do any testing/calculations about making sure we don't explode rabbitmq?

**Comment by timlesallen** [comment]
How do we make sure we start/stop at the right points... There is an argument for the offset to start at, is there one that tells us where to stop?

**Comment by timlesallen** [comment]
What is the procedure for knowing it has worked OK?  might be worth doing a github issue with what will be done and how it will be checked?

File: src/amqp-writable-stream.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+module.exports = class AmqpWritableStream extends Writable {
+  constructor (channel, queueName, opts = {}) {
+    super(Object.assign(opts, { objectMode: true }));
+    this.channel = channel;
+    this.queueName = queueName;
+    channel.on('drain', () => this.emit('drain'));
+  }
+  write (chunk, encoding, callback) {
+    return this.channel.sendToQueue(this.queueName, Buffer.from(stringify(chunk)));
+  }
```
Comment (timlesallen) [inline src/amqp-writable-stream.js:16]: Maybe some comments about how the backpressure works here would be good.  Does the `drain` event cause the backpressure to go away, coupled with the fact that we don't call `callback`?

**Comment by timlesallen** [comment]
hook up code climate properly on this PR?

**Pull request merged**